sdk/identity/azure-identity/azure/identity/_internal/msal_client.py

```diff
@@ -0,0 +1,137 @@
+                for secret in ("access_token", "refresh_token"):
+                    if secret in content:
+                        content[secret] = "***"
+                message = 'Unexpected response from Azure Active Directory: "{}"'.format(content)
+        else:
+            message = "Unexpected response from Azure Active Directory"
+
+        raise ClientAuthenticationError(message=message, response=self._response.http_response)
+
+
+class MsalClient(object):
+    """Wraps Pipeline according to msal.oauth2cli.http"""
+
+    def __init__(self, **kwargs):  # pylint:disable=missing-client-constructor-parameter-credential
+        # type: (**Any) -> None
+        self._pipeline = _build_pipeline(**kwargs)
+
+    def post(self, url, params=None, data=None, headers=None, **kwargs):  # pylint:disable=unused-argument
+        # type: (str, Optional[Dict[str, str]], RequestData, Optional[Dict[str, str]], **Any) -> MsalResponse
+        request = HttpRequest("POST", url, headers=headers)
+        if params:
+            request.format_parameters(params)
+        if data:
+            if isinstance(data, dict):
+                request.headers["Content-Type"] = "application/x-www-form-urlencoded"
+                request.set_formdata_body(data)
+            elif isinstance(data, six.text_type):
+                body_bytes = six.ensure_binary(data)
+                request.set_bytes_body(body_bytes)
+            else:
+                raise ValueError('expected "data" to be text or a dict')
+
+        response = self._pipeline.run(request)
+        return MsalResponse(response)
+
+    def get(self, url, params=None, headers=None, **kwargs):  # pylint:disable=unused-argument
+        # type: (str, Optional[Dict[str, str]], Optional[Dict[str, str]], **Any) -> MsalResponse
+        request = HttpRequest("GET", url, headers=headers)
+        if params:
+            request.format_parameters(params)
+        response = self._pipeline.run(request)
+        return MsalResponse(response)
+
+
+def _create_config(**kwargs):
+    # type: (Any) -> Configuration
+    config = Configuration(**kwargs)
+    config.logging_policy = NetworkTraceLoggingPolicy(**kwargs)
+    config.retry_policy = RetryPolicy(**kwargs)
+    config.proxy_policy = ProxyPolicy(**kwargs)
+    config.user_agent_policy = UserAgentPolicy(base_user_agent=USER_AGENT, **kwargs)
+    return config
+
+
+def _build_pipeline(config=None, policies=None, transport=None, **kwargs):
+    # type: (Optional[Configuration], Optional[PolicyList], Optional[HttpTransport], **Any) -> Pipeline
+    config = config or _create_config(**kwargs)
+
+    if policies is None:  # [] is a valid policy list
+        policies = [
+            ContentDecodePolicy(),
+            config.user_agent_policy,
+            config.proxy_policy,
+            config.retry_policy,
+            config.logging_policy,
+            DistributedTracingPolicy(**kwargs),
+            HttpLoggingPolicy(**kwargs),
+        ]
```

Comment:
Do we need header policy?

Reply:
No. User agent is the only header we want to set on every request.